### ISSUE TITLE
Can't upload source code for an add-on if the filename or slug contains non-ascii chars

### ISSUE THREAD

**Issue by panuscript** [mozilla/addons#132]
I don't know english well.
so, I explain with screen capture.

I upload my addon.
In Awaiting Review, A reviewer requested further information, and to upload source code.
but I cant upload source code, like this.

tried with windows7 64bit firefox, chrome, opera, And ubuntu 16.04 firefox, chrome, opera,
but result is always same thing.

![1](https://cloud.githubusercontent.com/assets/18691610/15809537/461d66c2-2bce-11e6-875a-9238d5ed0392.png)

![2](https://cloud.githubusercontent.com/assets/18691610/15809546/8e585e38-2bce-11e6-879c-5b99226312f8.png)

Oops! We had an error.

We'll get to fixing that soon.

You can try refreshing the page, or head back to the Add-ons homepage.

**Comment by EnTeQuAk** [comment]
This could be a duplicate of https://github.com/mozilla/addons-server/issues/2828

**Comment by andymckay** [comment]
It is, you can email the sources to amo-editors AT mozilla DOT org and I'll cc @wagnerand so he knows those sources are coming.

**Comment by wagnerand** [comment]
Are you sure it is a dupe of mozilla/addons-server#2828 ? That one is about uploading sources together with submitting a new version, which is not the case here according to the provided screenshot.

**Comment by andymckay** [comment]
No it's not, I think its: https://sentry.prod.mozaws.net/operations/olympia-prod/issues/250542/, which is a unicode error in the posix path, probably due to the filename of the add-on.

**Comment by diox** [comment]
I tried to write a test to reproduce this locally and failed. I did this:

```
 diff --git a/src/olympia/devhub/tests/test_views_versions.py b/src/olympia/devhub/tests/test_views_versions.py
 index ea80ba9..cf750e3 100644
 --- a/src/olympia/devhub/tests/test_views_versions.py
 +++ b/src/olympia/devhub/tests/test_views_versions.py
 @@ -1,3 +1,4 @@
 +# -*- coding: utf-8 -*-
  import re
  
  import mock
 @@ -553,6 +554,9 @@ class TestVersionEditDetails(TestVersionEditBase):
          assert doc('p.add-app')[0].attrib['class'] == 'add-app hide'
  
      def test_should_accept_zip_source_file(self):
 +        self.addon.update(slug=u'디시질질')
 +        self.url = reverse('devhub.versions.edit',
 +                           args=['디시질질', self.version.id])
          tdir = temp.gettempdir()
          tmp_file = temp.NamedTemporaryFile
          with tmp_file(suffix=".zip", dir=tdir) as source_file:
 @@ -564,6 +568,7 @@ class TestVersionEditDetails(TestVersionEditBase):
          version = Version.objects.get(pk=self.version.pk)
          assert version.source
          assert version.addon.admin_review
 +        assert version.source.storage.exists(version.source.path)
  
          # Check that the corresponding automatic activity log has been created.
          log = ActivityLog.objects.get(action=amo.LOG.REQUEST_SUPER_REVIEW.id)
```

And it passed. In that test, `version.source.path` is a `str`, and if I convert it to an unicode object, I can see the slug in here:
```
>>> self.addon.slug in version.source.path.decode('utf-8')
True
```

Which means django successfully converted the unicode path to a str before saving.

It's possible I made a mistake while testing, but I suspect this might be a configuration issue on our servers. The answers to [this stackoverflow question on the same topic](http://stackoverflow.com/questions/3715865/unicodeencodeerror-ascii-codec-cant-encode-character/7602446#7602446) seems to indicate we should configure this at the uwsgi level. Any ideas, @jasonthomas ?

Note: I suspect we don't have the issue in other places in our code because we either a) don't use a `FileField` and manually handle the file upload or b) set a path that don't end up containing unicode characters.

**Comment by jasonthomas** [comment]
Our nginx configuration already contains 'charset utf-8;', but LC_ALL and LC_LANG are not set. Are we able to replicate this issue on -dev?

**Comment by diox** [comment]
Yes, I just reproduced it on dev: https://sentry.prod.mozaws.net/operations/olympia-dev/issues/336084/

**Comment by jasonthomas** [comment]
We should test setting LANG and LC_ALL to 'en_US.UTF-8' on -dev uwsgi web workers and celery. @bqbn can you work on this?

**Comment by bqbn** [comment]
@diox the setting is done for both the web workers and celery on dev. Could you test it again to see if it still happens?

**Comment by diox** [comment]
Looks like it's still happening :(

https://sentry.prod.mozaws.net/operations/olympia-dev/issues/336831/

**Comment by jasonthomas** [comment]
Could you attach the add-on you are testing with so we can investigate further?

**Comment by panuscript** [comment]
[dczz326_full.zip](https://github.com/mozilla/addons/files/318835/dczz326_full.zip)

**Comment by bqbn** [comment]
I think the issue might be that https://github.com/mozilla/addons-server/blob/master/src/olympia/amo/utils.py#L827-L830 turned the unicode string
```
u'version_source/e/ne/None/about-uptime디시질질-0.3-src.zip'
```
into a byte code string (utf-8 encoded)
```
'version_source/e/ne/None/about-uptime디시질질-0.3-src.zip'
```
The error happens on https://github.com/mozilla/addons-server/blob/master/src/olympia/amo/utils.py#L824-L825 because it fails to concatenate self.location (a unicode string) and the return path name, which is a byte code string. 

Maybe we can try decoding the returned path name if six.PY3 is false. 
```
if six.PY3:
    return smart_str(string)
else:
    return smart_str(string).decode('utf-8')
```
```

**Comment by wagnerand** [comment]
I just tried this again and it fails no matter what filename (or contents) the source zip has. So, I think this is related to the name/slug of the add-on.
Not sure if that supports or contradicts @bqbn's comment.

**Comment by diox** [comment]
@wagnerand : you might be running into https://github.com/mozilla/addons-server/issues/2828 which is different.

@bqbn's might help this bug but I'm not so sure. I'd like to reproduce it locally (and have a test) first...

**Comment by wagnerand** [comment]
> @wagnerand : you might be running into mozilla/addons-server#2828 which is different.

Nope. I am not uploading a new version. I am trying to attach a dummy source to the add-on version in question on prod (https://addons.mozilla.org/en-US/editors/review/%D1%8F%D0%BD%D0%B4%D0%B5%D0%BA%D1%81-%D0%BA%D0%BE%D0%BB%D0%BB%D0%B5%D0%BA%D1%86%D0%B8%D0%B8) which fails still with the same error (https://sentry.prod.mozaws.net/operations/olympia-prod/issues/250542/)

**Comment by diox** [comment]
Ah. But I read your previous comment too quickly: we already know it will fail if _any_ component of the source filename on our servers (which depend on the add-on slug and the upload filename) contains non-ascii chars.

**Comment by wagnerand** [comment]
So, what does it take to fix this?

We advertise this feature to users and link to https://developer.mozilla.org/en-US/Add-ons/Source_Code_Submission#How_to_upload_source_code in our review messages.

**Comment by wagnerand** [comment]
Oh, now it was me who mixed up the issues. The previous comment was meant to go to https://github.com/mozilla/addons-server/issues/2828. Anyway, I think those two issues are related and fixing one might fix the other as well.

**Comment by wagnerand** [comment]
Adding @deadsquid to see whether we could re-prioritize this.

**Comment by ValentinaPC** [comment]
Not happening for add-ons without kanji chars in their names.

**Comment by panuscript** [comment]
I don't understand. 
I'm not using 'kanji'. (like 山江風)
i'm using korean. (like 산강풍)

**Comment by ValentinaPC** [comment]
@panuscript : I've used the "kanji" name as a _generic_ one - for all languages that are not using latin letters, that is why I put the quotation marks.

**Comment by panuscript** [comment]
oh, I got the point. I'm sorry to misunderstand.

but still, I have a question, I don't think that it's a solution - not using korean or something CJK language.
My extension is only for korean, so it's name must be a korean.

**Comment by diox** [comment]
It's not a solution. It describes the bug and what triggers it.

**Comment by ValentinaPC** [comment]
@panuscript : that is not the solution, is the cause :) 
Let's leave the people in charge to take care of it.

**Comment by panuscript** [comment]
oh, Sorry again for not having the patience.
I was too rushed. I will wait!  :)

**Comment by wagnerand** [comment]
I encountered this half a dozen times just this week. I know @diox is very busy, maybe someone else could have a look?

**Comment by diox** [comment]
I stand by what I said in https://github.com/mozilla/addons/issues/132#issuecomment-224585786, I still think it's not a code issue. I've tried it again locally, with an add-on that had the same kind of non-ascii slug & name as the one from the description of this issue, with a source filename that also has these chars, and it worked fine. The same thing fails on dev/stage/prod.

**Comment by wagnerand** [comment]
@diox are you saying it must be an environment issue? How do we figure this out?

**Comment by diox** [comment]
I'm saying that I _suspect_ it's an environment issue, yes. I've basically been retracing my steps from https://github.com/mozilla/addons/issues/132#issuecomment-224585786 and reached the exact same conclusions. The only way forward that I see is to make tweaks to the environments we know we can reproduce the bug in.

**Comment by wagnerand** [comment]
@jasonthomas is there a way for developers to get their local environment closer to what runs on prod?

**Comment by jasonthomas** [comment]
We do a few things differently in our deployment environment. I believe the biggest difference is using uWSGI as our WSGI server.

I would try using uwsgi with https://github.com/mozilla/addons-server/blob/master/src/olympia/wsgi.py instead of runserver. You can check out the current uwsgi configuration from the dev server at `/etc/uwsgi.d/addons.ini`.

**Comment by wagnerand** [comment]
@diox any updates on this. Can I help figuring this out?

**Comment by diox** [comment]
I haven't had time to get back on this myself.

**Comment by EnTeQuAk** [comment]
Looking into this partially currently because of a few other issues that appear to be filename encoding related.

**Comment by muffinresearch** [comment]
@EnTeQuAk was this fixed with the other recent unicode fixes?

**Comment by ValentinaPC** [comment]
- I've tried to upload a source file from an add-on with non-ascii chars in its name and worked with success.
- I've tried uploading a source code file while being in `ko` language
- I've submitted an add-on using Korean locale and added a source code with success.

From my point of view, this was fixed, meanwhile.
Video: 
![nonascii](https://cloud.githubusercontent.com/assets/15685960/21185014/a43b8efc-c217-11e6-9f22-91280e6ce430.gif)
